scripts/builtin/decisionTree.dml

```diff
@@ -0,0 +1,449 @@
+            create_child_nodes_flag = impurity < as.scalar(parent_impurity)
+            print("Current impurity:   " + impurity)
+            printVector("Current threshold:  ", threshold) #Todo: threshold can't be printed if categorical
+        }
+        print("Current column:     " + used_col)
+        print("Current type:       " + type)
+        if (create_child_nodes_flag) {
+            [left, right] = calculateChildNodes(node)
+            node_queue = dataQueuePush(left, right, node_queue)
+
+            [new_use_cols_vector, left_use_rows_vector, right_use_rows_vector] = splitData(X, use_rows_vector, use_cols_vector, used_col, threshold, type)
+            use_rows_queue = dataQueuePush(left_use_rows_vector, right_use_rows_vector, use_rows_queue)
+            use_cols_queue = dataQueuePush(new_use_cols_vector, new_use_cols_vector, use_cols_queue)
+
+            impurity_queue = dataQueuePush(matrix(impurity, rows = 1, cols = 1), matrix(impurity, rows = 1, cols = 1), impurity_queue)
+            offset = dataQueueLength(node_queue) - 1
+            print("Offset to left child: " + offset)
+            M = outputMatrixBind(M, node, offset, used_col, R, threshold)
+        } else {
+            M = outputMatrixBind(M, node, 0.0, used_col, R, matrix(0, rows = 1, cols = 1))
+            print("LEAF!")
+        }
+        queue_length = dataQueueLength(node_queue)# -- user-defined function calls not supported in relational expressions
+
+        print("New QueueLen:       " + queue_length)
+        print("-------------------------------------------------------------------------------------------------------")
+        print(" ")
+    }
+}
+
+# --------------------- Missuses matrix as a queue for vectors ---------------------------------------------------------
+dataQueueLength = function(Matrix[Double] queue)  return (Double len) {
+    len = ncol(queue)
+}
+
+dataQueuePop = function(Matrix[Double] queue)  return (Matrix[Double] new_queue, Matrix[Double] node) {
+    node = matrix(queue[,1], rows=1, cols=nrow(queue))     # reshape to force the creation of a new object
+    node = matrix(node, rows=nrow(queue), cols=1)          # reshape to force the creation of a new object
+    len = dataQueueLength(queue)
+    if (len < 2) {
+        new_queue = matrix(0,0,0)
+    } else {
+        new_queue = matrix(queue[,2:ncol(queue)], rows=nrow(queue), cols=ncol(queue)-1)
+    }
+}
+
+dataQueuePush = function(Matrix[Double] left, Matrix[Double] right, Matrix[Double] queue)  return (Matrix[Double] new_queue) {
+    len = dataQueueLength(queue)
+    if(len <= 0) {
+        new_queue = cbind(left, right)
+    } else {
+        new_queue = cbind(queue, left, right)
+    }
+}
+
+#-----------------------------------------------------------------------------------------------------------------------
+# --------------------- Missuses matrix as a vectors for Doubles -------------------------------------------------------
```

Comment:
i would not call using matrix as vectors a misuse, under the covers all matrices are vectors in systemds.